Week4/HW04.hs

```diff
@@ -0,0 +1,149 @@
+                withoutTrailingZeros = dropWhileEnd (==0) p
+
+-- Exercise 4 -----------------------------------------
+
+plus :: Num a               => Poly a -> Poly a -> Poly a
+plus (P polyA) (P polyB)    = P $ zipWith (+) (polyA ++ replicate leadingZerosToA 0) (polyB ++ replicate leadingZerosToB 0)
+    where 
+        lA                  = length polyA 
+        lB                  = length polyB
+        leadingZerosToA 
+            | lA < lB       = lB - lA
+            | otherwise     = 0
+        leadingZerosToB 
+            | lB < lA       = lA - lB
+            | otherwise     = 0
+
+-- Exercise 5 -----------------------------------------
+
+-- c is the coeff being multiplied from the first polynomial
+-- e is the exponentioal term being used for multiplication from first polynomial
+
+-- timesHelper (P [1, 4, 2]) 1 0 == P [1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 1 1 == P [0, 1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 1 2 == P [0, 0, 1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 2 2 == P [0, 0, 2, 8, 4]
+
+-- 1. add leading zeros list of size == e 
+-- and multiply rest with c
+timesHelper ::  (Num a)     => Poly a -> a -> Int -> Poly a
+timesHelper (P p) c e       = P $ replicate e 0 ++ map (*c) p
+
+-- getAllTimes 0 (P [1, 1, 2]) (P [1, 4, 2]) == [P [1, 4, 2], P [0, 1, 4, 2], P [0, 0, 2, 8, 4]]
+getAllTimes ::  (Num a)     => Int -> Poly a -> Poly a -> [Poly a]
+getAllTimes _ (P []) _      = []
+getAllTimes n (P [p]) q     = [timesHelper q p n]
+getAllTimes n (P (p:ps)) q  = timesHelper q p n :  getAllTimes (n+1) (P ps) q
+
+-- addAllTimes [P [1, 4, 2], P [0, 1, 4, 2], P [0, 0, 2, 8, 4]] == P [1, 5, 8, 10, 4]
+-- This is how it works with the help of [timesHelper] function
+--      [1, 4, 2, 0, 0]
+--   +  [0, 1, 4, 2, 0]
+--   +  [0, 0, 2, 8, 4]
+--      ---------------
+--      [1, 5, 8, 10, 4]
+--      ---------------
+addAllTimes :: (Num a)      => [Poly a] -> Poly a
+addAllTimes []              = P []
+addAllTimes [p1]            = p1
+addAllTimes [p1, p2]        = plus p1 p2
+addAllTimes (p1:p2:ps)      = plus p1 p2 + addAllTimes ps
+
+times :: (Num a)            => Poly a -> Poly a -> Poly a
+times p q                   = addAllTimes $ getAllTimes 0 p q
+
+-- Exercise 6 -----------------------------------------
+
+instance Num a             => Num (Poly a) where
+    (+)                    = plus
+    (*)                    = times
+    negate                 = (* P [-1])
+    fromInteger i          = P [fromInteger i]
+
+    -- No meaningful definitions exist
+    abs    = undefined
+    signum = undefined
+
+-- Exercise 7 -----------------------------------------
+
+applyP :: Num a            => Poly a -> a -> a
+applyP (P p) valX          = sum $ zipWith (\a b -> a * (valX ^ b)) p intList
+    where 
+        intList :: [Int]
+        intList            = [0..]
+
+-- Exercise 8 -----------------------------------------
+
+-- I wasnt able to understand what is nderiv here so this is the only exc left in this weeks assignment 
```

Comment:
I think it's n-th derivative

so e.g. 
```haskell
nderiv 1 == deriv
nderiv 2 "x^3" == nderiv 1 "3*x^2" == "6*x"
```

Reply:
This could be derived from deriv recursively, this is what the assignment asked for i believe
```
nderiv n 
        | n <= 0            = id
        | otherwise         = deriv . nderiv (n-1) 
```

Reply:
Yes